[3.7]BL-4307 Change UI in Collection Chooser

gmartin7 opened the pull request:
Extracted a common SetupUiMenu method

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/bloombooks/bloomdesktop/1522)
<!-- Reviewable:end -->

andrew-polk commented:
Reviewed 6 of 7 files at r1.
Review status: 6 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[DistFiles/localization/Bloom.en.tmx, line 2732 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Kdzy2cZxCWl8866M_Vy:-Kdzy2cZxCWl8866M_Vz:bemyuh) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/DistFiles/localization/Bloom.en.tmx#L2732)):*
> <seg></seg>

Adding this with an empty seg element was unintentional, right?
Probably needs to have the L10NSharp priority set to NotLocalizable.

---

*[src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs, line 38 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-KdzzHk3DGmYyZBJ8U7U:-KdzzHk3DGmYyZBJ8U7V:b-pny8ch) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs#L38)):*
> const int fiddle = 8;

Not sure how much we care, but I assume this will break if a user has their default font size set differently than yours in Windows. I wouldn't be shocked if it needed to be different in Linux, too.

---

*[src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs, line 42 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke--olJontOxOXLOz-P:-Ke--olJontOxOXLOz-Q:b-uymlwc) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs#L42)):*
> ```Smalltalk
> 
> 			_uiLanguageMenu.DropDownItems.Clear();
> 			UiMenuRenderer.SetupUiMenu(_uiLanguageMenu, (sender, args) =>
> ```

May not be worth it, but it would have helped me a lot to understand what was going on here if this was actually a named delegate. (with a name like UiLanguageMenuItemClickHandler)
Same for other call.

---

*[src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs, line 112 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-KdzylHG6215_jlNdZ5R:-KdzylHG6215_jlNdZ5S:b-v07lfu) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs#L112)):*
> // We've pulled label1 out into _topRightPanel; do the same check·

There's a trailing space on this line. Do you have the editorconfig extension installed in VS?

---

*[src/BloomExe/MiscUI/UiMenuRenderer.cs, line 15 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-0B-IPh2cEOrIFEHk:-Ke-0B-IPh2cEOrIFEHl:b-iva1ni) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/MiscUI/UiMenuRenderer.cs#L15)):*
> UiMenuRenderer

I think this would be more clear if named UiLanguageMenuRenderer. Same with method name.

---

*[src/BloomExe/MiscUI/UiMenuRenderer.cs, line 15 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-1qPLRco1JbivLGnB:-Ke-1qPLRco1JbivLGnC:b-j2si0i) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/MiscUI/UiMenuRenderer.cs#L15)):*
> ```Smalltalk
> 	/// http://stackoverflow.com/questions/23013481/toolstrip-drop-down-button-not-large-enough
> 	/// </summary>
> 	public class UiMenuRenderer : ToolStripProfessionalRenderer
> ```

Is this class needed if we aren't making the dropdown arrow larger?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522)*
<!-- Sent from Reviewable.io -->

JohnThomson commented:
Good start, but more could be in the common method...and I'm not clear on the reasons for the differences in how choosing a UI language is handled. See individual comments.

---

Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[DistFiles/localization/Bloom.en.tmx, line 2732 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-KdzzI4rWgWwBc43w9tt:-KdzzI4s5_gngG2ZHNXw:b-mfy2t) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/DistFiles/localization/Bloom.en.tmx#L2732)):*
> ```
>     <tu tuid="OpenCreateNewCollectionsDialog.UILanguageMenu">
>       <tuv xml:lang="en">
>         <seg></seg>
> ```

Why is English empty?

---

*[src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs, line 44 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-0cAtTAn5HkS8EYQM:-Ke-0cAtTAn5HkS8EYQN:b-x3xz0c) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs#L44)):*
> ```Smalltalk
> 			UiMenuRenderer.SetupUiMenu(_uiLanguageMenu, (sender, args) =>
> 			{
> 				var item = sender as ToolStripItem;
> ```

Difficult to follow this block, possibly because I haven't read SetupUiMenu to know how it invokes this block. A comment would be helpful, along the lines "This block is invoked when the user selects a UI language. sender is the menu item for the particular language. It's responsible to set the UI language and update the size of...
At another level, I wonder whether it would be possible/desirable for SetupUIMenu to arrange for some of this behavior...presumably in all contexts selecting one of these items should SetUILanguage?

---

*[src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs, line 46 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Kdzzi_bZF0dTKiMQLkc:-Kdzzi_bZF0dTKiMQLkd:b-ck3g0q) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs#L46)):*
> ```Smalltalk
> 				var item = sender as ToolStripItem;
> 				if (item == null)
> 					return; // shouldn't happen
> ```

If it shouldn't happen, should we ignore it? If not, consider declaring the class of sender and simply casting item.Tag.

---

*[src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs, line 51 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-1EiMyQ08woadonSS:-Ke-1EiMyQ08woadonST:b663vc0) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs#L51)):*
> ```Smalltalk
> 					return; // shouldn't happen
> 
> 				_toolStrip1.Width = 100;
> ```

Comment explaining why this is necessary, please...also if it's needed here, why not everywhere this UI menu is used?

---

*[src/BloomExe/MiscUI/UiMenuRenderer.cs, line 13 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke--j-W00_ulp0m7gPq:-Ke--j-XNT_1wctu1Zb3:bxc8ky1) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/MiscUI/UiMenuRenderer.cs#L13)):*
> ```Smalltalk
> 	/// <summary>
> 	/// The idea for this class is from:
> 	/// http://stackoverflow.com/questions/23013481/toolstrip-drop-down-button-not-large-enough
> ```

This history is nice to have, but I'd still like to see a summary of the class's purpose. Is the whole thing superfluous if JohnH continues to prefer the smaller arrow?

---

*[src/BloomExe/MiscUI/UiMenuRenderer.cs, line 75 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-2336c55xNJEa1QlA:-Ke-2336c55xNJEa1QlB:b-f26eo6) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/MiscUI/UiMenuRenderer.cs#L75)):*
> ```Smalltalk
> 		//}
> 
> 		public static void SetupUiMenu(ToolStripDropDownButton uiMenuControl, EventHandler clickHandler)
> ```

It feels strange that the rest of this class is all about rendering a menu, while this is about Localization. I suppose it makes sense not to split them apart until we have another menu that needs the same tweaking, but the class comment should point out why the two responsibilities are being combined and suggest splitting if appropriate.

---

*[src/BloomExe/MiscUI/UiMenuRenderer.cs, line 88 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-2iW8GAET1VUyjtKZ:-Ke-2iW9HKsVYrYtY03w:baj7tw3) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/MiscUI/UiMenuRenderer.cs#L88)):*
> ```Smalltalk
> 				var item = uiMenuControl.DropDownItems.Add(lang.NativeName + englishName);
> 				item.Tag = lang;
> 				var itemNativeName = ((CultureInfo)item.Tag).NativeName;
> ```

That is, itemNativeName = lang.NativeName? So why not the short form? Why even have another variable? It's only used once.

---

*[src/BloomExe/Workspace/WorkspaceView.cs, line 291 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-3H1H2OGcFuwt5MW6:-Ke-3H1H2OGcFuwt5MW7:b4hc99d) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/Workspace/WorkspaceView.cs#L291)):*
> ```Smalltalk
> 			UiMenuRenderer.SetupUiMenu(_uiLanguageMenu, (sender, args) =>
> 			{
> 				var item = sender as ToolStripItem;
> ```

Duplicates much of what the other client does. Can you pull the common stuff into the common method (as a clickHandler block there) and make the second argument of this method an Action giving just the info the clients need?

---

*[src/BloomExe/Workspace/WorkspaceView.cs, line 304 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-3omCIFkOXMANyocY:-Ke-3omCIFkOXMANyocZ:bokfhxo) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/Workspace/WorkspaceView.cs#L304)):*
> ```Smalltalk
> 				item.Select();
> 				_uiLanguageMenu.Text = ((CultureInfo)item.Tag).NativeName;
> 				SaveOriginalButtonTexts();
> ```

Why is this needed here but not in the other context? Also other differences between the two.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-:-Ke-46SyGnyHIOH_c7rO:bk7x11v)*
<!-- Sent from Reviewable.io -->

gmartin7 commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[DistFiles/localization/Bloom.en.tmx, line 2732 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Kdzy2cZxCWl8866M_Vy:-Ke9y4tiWn8CblZxVMg4:brpa7m6) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/DistFiles/localization/Bloom.en.tmx#L2732)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> <seg></seg>

Adding this with an empty seg element was unintentional, right?
Probably needs to have the L10NSharp priority set to NotLocalizable.
</blockquote></details>

I need to move on to FLEx now, but I can answer this one. It may not be necessary to have this one at all. The one that's needed is the following "OpenCreateNewCollectionsDialog.UILanguageMenu-tooltip". I wasn't sure how L10NSharp dealt with the tooltip feature, and I haven't seen a 'x-tooltip' entry without the corresponding 'parent' entry.

---

*[src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs, line 51 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-1EiMyQ08woadonSS:-Ke9ymqZ6A1kPCG8ezW2:b-4o0kia) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs#L51)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Comment explaining why this is necessary, please...also if it's needed here, why not everywhere this UI menu is used?
</blockquote></details>

Hmm... maybe comment it out and see if it still works? Maybe it's not necessary... I'm not sure.

---

*[src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs, line 112 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-KdzylHG6215_jlNdZ5R:-Ke9yxv1FXLa5i_zbfXu:b-ej0tyu) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs#L112)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> // We've pulled label1 out into _topRightPanel; do the same check·

There's a trailing space on this line. Do you have the editorconfig extension installed in VS?
</blockquote></details>

Oops. I know I have it in VSCode, but perhaps not in VS.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522)*
<!-- Sent from Reviewable.io -->

gmartin7 commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[DistFiles/localization/Bloom.en.tmx, line 2732 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-KdzzI4rWgWwBc43w9tt:-KeB4w3o6sj4nHvMBc-P:bt2vr1b) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/DistFiles/localization/Bloom.en.tmx#L2732)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Why is English empty?
</blockquote></details>

Fixed.

---

*[DistFiles/localization/Bloom.en.tmx, line 2732 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Kdzy2cZxCWl8866M_Vy:-KeB4z1lZD_7_KQADUOI:b-u6xa4b) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/DistFiles/localization/Bloom.en.tmx#L2732)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary><blockquote>

I need to move on to FLEx now, but I can answer this one. It may not be necessary to have this one at all. The one that's needed is the following "OpenCreateNewCollectionsDialog.UILanguageMenu-tooltip". I wasn't sure how L10NSharp dealt with the tooltip feature, and I haven't seen a 'x-tooltip' entry without the corresponding 'parent' entry.
</blockquote></details>

Since fixed.

---

*[src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs, line 38 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-KdzzHk3DGmYyZBJ8U7U:-KeApxLTH6vnTGaerTKx:b-gurx66) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs#L38)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> const int fiddle = 8;

Not sure how much we care, but I assume this will break if a user has their default font size set differently than yours in Windows. I wouldn't be shocked if it needed to be different in Linux, too.
</blockquote></details>

Could be. The problem is that the Form has large icons for the button opposite and the text is centered vertically. I couldn't figure out any other way to line up the text the way JohnH wanted it.

---

*[src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs, line 42 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke--olJontOxOXLOz-P:-KeB9S9gW-kcmTCecAUX:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs#L42)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

May not be worth it, but it would have helped me a lot to understand what was going on here if this was actually a named delegate. (with a name like UiLanguageMenuItemClickHandler)
Same for other call.
</blockquote></details>

Done.

---

*[src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs, line 44 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-0cAtTAn5HkS8EYQM:-KeB9dtL6WECbi5hMr6U:bwj5er) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs#L44)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Difficult to follow this block, possibly because I haven't read SetupUiMenu to know how it invokes this block. A comment would be helpful, along the lines "This block is invoked when the user selects a UI language. sender is the menu item for the particular language. It's responsible to set the UI language and update the size of...
At another level, I wonder whether it would be possible/desirable for SetupUIMenu to arrange for some of this behavior...presumably in all contexts selecting one of these items should SetUILanguage?
</blockquote></details>

Hmm... is it clearer now?

---

*[src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs, line 46 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Kdzzi_bZF0dTKiMQLkc:-KeAv1SCGgq7W9uu6-pq:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs#L46)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

If it shouldn't happen, should we ignore it? If not, consider declaring the class of sender and simply casting item.Tag.
</blockquote></details>

Done.

---

*[src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs, line 51 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-1EiMyQ08woadonSS:-KeAvJBAUB7aTG_5e8Uk:b-ezv303) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs#L51)):*
> comment it out and see if it still works?
Nope. It doesn't work. I commented in the code. Not sure why other menus work w/o it.

---

*[src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs, line 112 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-KdzylHG6215_jlNdZ5R:-KeAviZg3Q0Av3OwLFBj:b-uv0nz9) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs#L112)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary><blockquote>

Oops. I know I have it in VSCode, but perhaps not in VS.
</blockquote></details>

Umm... actually, yes, I do have it. This is 3.7, tho'. Does that make a difference? Removed the space.

---

*[src/BloomExe/MiscUI/UiMenuRenderer.cs, line 13 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke--j-W00_ulp0m7gPq:-KeB-_AUG4K6bbG-kQKk:bfxrvvv) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/MiscUI/UiMenuRenderer.cs#L13)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

This history is nice to have, but I'd still like to see a summary of the class's purpose. Is the whole thing superfluous if JohnH continues to prefer the smaller arrow?
</blockquote></details>

Summary given. Other parts look better with this renderer too. See comments added to the class.

---

*[src/BloomExe/MiscUI/UiMenuRenderer.cs, line 15 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-1qPLRco1JbivLGnB:-KeAy6paOQjnH7EVAkwD:b-1i7p0i) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/MiscUI/UiMenuRenderer.cs#L15)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Is this class needed if we aren't making the dropdown arrow larger?
</blockquote></details>

I didn't know about Workspace.NoBorderToolStripRenderer until late in the process. So if JH is happy with the blue selection backcolor of that class, yes, we could use it. The main difference (besides the larger dropdown arrow possibility) is the matching gray selection backcolor of the menu items; it's more like the other buttons' hover backcolor on the OpenCreateCloneControl. Or I can pull it out and use NoBorderToolStripRenderer. It does introduce a dependency on Workspace... or we could even pull the NoBorderToolStripRenderer out to MiscUI. Right now, WorkspaceView and LibraryView use the NoBorderToolStripRenderer.

---

*[src/BloomExe/MiscUI/UiMenuRenderer.cs, line 15 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-0B-IPh2cEOrIFEHk:-KeAz13P_zThYAQO0rlZ:b-hff6wv) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/MiscUI/UiMenuRenderer.cs#L15)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> UiMenuRenderer

I think this would be more clear if named UiLanguageMenuRenderer. Same with method name.
</blockquote></details>

renamed

---

*[src/BloomExe/MiscUI/UiMenuRenderer.cs, line 75 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-2336c55xNJEa1QlA:-KeAwDsLPVOv9LW4O8-K:b-bb8c2h) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/MiscUI/UiMenuRenderer.cs#L75)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

It feels strange that the rest of this class is all about rendering a menu, while this is about Localization. I suppose it makes sense not to split them apart until we have another menu that needs the same tweaking, but the class comment should point out why the two responsibilities are being combined and suggest splitting if appropriate.
</blockquote></details>

Yes, I couldn't find a better place to put it. Suggestions?

---

*[src/BloomExe/MiscUI/UiMenuRenderer.cs, line 88 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-2iW8GAET1VUyjtKZ:-KeB87zQu64_mc_ykk-Q:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/MiscUI/UiMenuRenderer.cs#L88)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

That is, itemNativeName = lang.NativeName? So why not the short form? Why even have another variable? It's only used once.
</blockquote></details>

Done.

---

*[src/BloomExe/Workspace/WorkspaceView.cs, line 291 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-3H1H2OGcFuwt5MW6:-KeB9W4f_PS6W0Ns3t9J:bbeugn9) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/Workspace/WorkspaceView.cs#L291)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Duplicates much of what the other client does. Can you pull the common stuff into the common method (as a clickHandler block there) and make the second argument of this method an Action giving just the info the clients need?
</blockquote></details>

Might have to leave this as an exercise for you guys....

---

*[src/BloomExe/Workspace/WorkspaceView.cs, line 304 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-3omCIFkOXMANyocY:-KeB8yD2NFxx0Fg4Kgfp:b-z7301s) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/Workspace/WorkspaceView.cs#L304)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Why is this needed here but not in the other context? Also other differences between the two.
</blockquote></details>

This is for the WorkspaceView usage where the user may shrink the window, causing the various buttons to resize to fit and reduce the amount of text they bear, etc. Perhaps part of what they do here also makes the resizing of the item that I do in the other version unnecessary, but getting these methods to do something in the other context doesn't really make sense.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522)*
<!-- Sent from Reviewable.io -->

andrew-polk commented:
Since this is targeting 3.7, we must be especially confident in the solution.

---

Reviewed 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs, line 38 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-KdzzHk3DGmYyZBJ8U7U:-KeE63nZwEsnyDS-Qkm7:bekh46g) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs#L38)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary><blockquote>

Could be. The problem is that the Form has large icons for the button opposite and the text is centered vertically. I couldn't figure out any other way to line up the text the way JohnH wanted it.
</blockquote></details>

I'm not comfortable with this yet. We at least need a test to see how it behaves if user has Windows font set to 125% for example.

---

*[src/BloomExe/MiscUI/UiLanguageMenuRenderer.cs, line 15 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-1qPLRco1JbivLGnB:-KeE6h8mlPai7TppKudZ:bb2bkvv) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/MiscUI/UiLanguageMenuRenderer.cs#L15)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary><blockquote>

I didn't know about Workspace.NoBorderToolStripRenderer until late in the process. So if JH is happy with the blue selection backcolor of that class, yes, we could use it. The main difference (besides the larger dropdown arrow possibility) is the matching gray selection backcolor of the menu items; it's more like the other buttons' hover backcolor on the OpenCreateCloneControl. Or I can pull it out and use NoBorderToolStripRenderer. It does introduce a dependency on Workspace... or we could even pull the NoBorderToolStripRenderer out to MiscUI. Right now, WorkspaceView and LibraryView use the NoBorderToolStripRenderer.
</blockquote></details>

I don't know anything about the other renderer. But if we can use it and keep from adding a new class, let's do it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-:-KeE5st7al9nPQaW5g1P:b-oa57rs)*
<!-- Sent from Reviewable.io -->

JohnThomson commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs, line 51 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-1EiMyQ08woadonSS:-KeF49mIVGr2u-YPtfoQ:bq735us) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs#L51)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary><blockquote>

> comment it out and see if it still works?
Nope. It doesn't work. I commented in the code. Not sure why other menus work w/o it.
</blockquote></details>

And I guess we can't just set it right away to _uiLanguageMenu.Width + 1 because that won't give the right answer until AFTER we set  _uiLanguageMenu.Text ? OK, I guess this can stand...I'd like to understand why it collapses better, but at least you've said what to watch out for if the line seems superfluous.

---

*[src/BloomExe/MiscUI/UiLanguageMenuRenderer.cs, line 15 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-1qPLRco1JbivLGnB:-KeF7gRcaIxfosX8Rqy5:b-gypar0) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/MiscUI/UiLanguageMenuRenderer.cs#L15)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

I don't know anything about the other renderer. But if we can use it and keep from adding a new class, let's do it.
</blockquote></details>

Especially for 3.7. If we need to we can tune things up in 3.9.

---

*[src/BloomExe/MiscUI/UiLanguageMenuRenderer.cs, line 75 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-2336c55xNJEa1QlA:-KeF5NkxpIKSnSkvlzJj:bw81wuy) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/MiscUI/UiLanguageMenuRenderer.cs#L75)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary><blockquote>

Yes, I couldn't find a better place to put it. Suggestions?
</blockquote></details>

I think it would be more at home as a public static method of WorkspaceView. That's the main place this menu occurs; it's something of an oddity that we also need it in the collection chooser. It certainly doesn't belong in a class which claims to be a generic way of tweaking a menu's appearance.

---

*[src/BloomExe/Workspace/WorkspaceView.cs, line 291 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-3H1H2OGcFuwt5MW6:-KeF5zvndp6297hLA96L:b-m9f2yv) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/Workspace/WorkspaceView.cs#L291)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary><blockquote>

Might have to leave this as an exercise for you guys....
</blockquote></details>

Minimal changes to what we already have feels good for a 3.7 patch, especially.  If we head in the direction I suggested of letting the main responsibility for setting up the UI language menu remain in WorkspaceView, maybe we could just break out any code from here that we want to share with the collection chooser into public static methods? In the process it might become clearer that we hadn't made any substantive changes to the existing code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522)*
<!-- Sent from Reviewable.io -->

gmartin7 commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs, line 38 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-KdzzHk3DGmYyZBJ8U7U:-KeEwsD9BBW7HXd9Z8Px:b-66n0k3) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/CollectionChoosing/OpenCreateCloneControl.cs#L38)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

I'm not comfortable with this yet. We at least need a test to see how it behaves if user has Windows font set to 125% for example.
</blockquote></details>

Tested at 125% and at 150% and surprisingly it still lines up!

---

*[src/BloomExe/MiscUI/UiLanguageMenuRenderer.cs, line 15 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-1qPLRco1JbivLGnB:-KeEwjblobqyb9ZmEe8P:b-fb5xu5) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/MiscUI/UiLanguageMenuRenderer.cs#L15)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

I don't know anything about the other renderer. But if we can use it and keep from adding a new class, let's do it.
</blockquote></details>

Will use NoBorderToolStripRenderer. Of course, now I HAVE to find a better place to put the common 'SetupUiMenu()' method. :(

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522)*
<!-- Sent from Reviewable.io -->

gmartin7 commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[src/BloomExe/MiscUI/UiLanguageMenuRenderer.cs, line 75 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-2336c55xNJEa1QlA:-KeFBWPZIzsxqc4lbc1H:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/MiscUI/UiLanguageMenuRenderer.cs#L75)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

I think it would be more at home as a public static method of WorkspaceView. That's the main place this menu occurs; it's something of an oddity that we also need it in the collection chooser. It certainly doesn't belong in a class which claims to be a generic way of tweaking a menu's appearance.
</blockquote></details>

Done.

---

*[src/BloomExe/Workspace/WorkspaceView.cs, line 291 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-Ke-3H1H2OGcFuwt5MW6:-KeFBc0TZbCs_vkpyOMl:b-3jra86) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/d2cd9cea6032f65c995096813ff73c36ca44a98a/src/BloomExe/Workspace/WorkspaceView.cs#L291)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Minimal changes to what we already have feels good for a 3.7 patch, especially.  If we head in the direction I suggested of letting the main responsibility for setting up the UI language menu remain in WorkspaceView, maybe we could just break out any code from here that we want to share with the collection chooser into public static methods? In the process it might become clearer that we hadn't made any substantive changes to the existing code.
</blockquote></details>

Have fun! :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522)*
<!-- Sent from Reviewable.io -->

andrew-polk commented:
I'll see what I can do...

---

Reviewed 1 of 7 files at r2, 4 of 5 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1522#-:-KeFh7RX7MUFU3WvlyP4:b34286i)*
<!-- Sent from Reviewable.io -->

andrew-polk commented:
Superceded by https://github.com/BloomBooks/BloomDesktop/pull/1542. I don't know how to add to this PR since it is on another person's fork. Starting fresh will likely be easier to review, anyway.
